Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDataDetectorOverlayController.mm

```diff
@@ -0,0 +1,292 @@
+    auto selectionBoundsInPluginSpace = plugin->rectForSelectionInPluginSpace(dataDetectorItem.selection().get());
+    auto visibleContentRectInPluginSpace = plugin->visibleContentRect();
+
+    return ::WebKit::createPlatformDataDetectorHighlight(Vector<FloatRect>::from(WTFMove(selectionBoundsInPluginSpace)), WTFMove(visibleContentRectInPluginSpace));
+}
+
+bool PDFDataDetectorOverlayController::handleMouseEvent(const WebMouseEvent& event, PDFDocumentLayout::PageIndex pageIndex)
+{
+    RefPtr plugin = protectedPlugin();
+    if (!plugin)
+        return false;
+
+    if (!PAL::isDataDetectorsFrameworkAvailable())
+        return false;
+
+    updateDataDetectorHighlightsIfNeeded(pageIndex);
+
+    installOverlayIfNeeded();
+
+    auto mousePositionInPluginSpace = plugin->mousePositionInView(event);
+    bool mouseIsOverActiveHighlightButton = false;
+    RefPtr previousActiveHighlight = std::exchange(m_activeDataDetectorHighlight, nullptr);
+    RefPtr<PDFDataDetectorItem> dataDetectorItemForActiveHighlight;
+
+    if (auto iterator = m_pdfDataDetectorItemsWithHighlightsMap.find(pageIndex); iterator != m_pdfDataDetectorItemsWithHighlightsMap.end()) {
+        for (auto& [dataDetectorItem, coreHighlight] : iterator->value) {
+            Boolean isOverButton = NO;
+            if (!PAL::softLink_DataDetectors_DDHighlightPointIsOnHighlight(coreHighlight->highlight(), mousePositionInPluginSpace, &isOverButton))
+                continue;
+
+            mouseIsOverActiveHighlightButton = isOverButton;
+            m_activeDataDetectorHighlight = coreHighlight.copyRef();
+            dataDetectorItemForActiveHighlight = dataDetectorItem.copyRef();
+            break;
+        }
+    }
+
+    if (previousActiveHighlight != m_activeDataDetectorHighlight) {
+        RefPtr overlay = protectedOverlay();
+
+        if (previousActiveHighlight)
+            previousActiveHighlight->fadeOut();
+
+        if (m_activeDataDetectorHighlight) {
+            overlay->layer().addChild(m_activeDataDetectorHighlight->layer());
+            m_activeDataDetectorHighlight->fadeIn();
+        }
+
+        overlay->setNeedsDisplay();
+        didInvalidateHighlightOverlayRects(ShouldUpdatePlatformHighlightData::No);
+    }
+
+    if (event.type() == WebEventType::MouseDown && mouseIsOverActiveHighlightButton)
+        return handleDataDetectorAction(WTFMove(mousePositionInPluginSpace), *dataDetectorItemForActiveHighlight);
+
+    return false;
+}
+
+bool PDFDataDetectorOverlayController::handleDataDetectorAction(IntPoint&& mousePositionInPluginSpace, PDFDataDetectorItem& dataDetectorItem)
+{
+    RefPtr plugin = protectedPlugin();
+
+    if (!plugin)
+        return false;
+
+    auto rectForSelectionInPluginSpace = roundedIntRect(plugin->rectForSelectionInPluginSpace(dataDetectorItem.selection().get()));
+
+    plugin->handleClickForDataDetectionResult({ dataDetectorItem.scannerResult(), WTFMove(rectForSelectionInPluginSpace) }, WTFMove(mousePositionInPluginSpace));
```

Comment:
Dunno if it's worth moving rects and points.

Reply:
Yes, I suppose it doesn't really help. I'll remove the moves.